components/resolved_ts/src/advance.rs

```diff
@@ -136,68 +143,54 @@ impl<E: KvEngine> AdvanceTsWorker<E> {
         // region_id -> peers id, record the responses
         let mut resp_map: HashMap<u64, Vec<u64>> = HashMap::default();
-        let mut leader_info_size = 0;
-        {
-            let meta = store_meta.lock().unwrap();
-            let store_id = match meta.store_id {
-                Some(id) => id,
-                None => return vec![],
-            };
-            for region_id in regions {
-                if let Some(region) = meta.regions.get(&region_id) {
-                    if let Some((term, leader_id)) = meta.leaders.get(&region_id) {
-                        let leader_store_id = find_store_id(&region, *leader_id);
-                        if leader_store_id.is_none() {
-                            continue;
-                        }
-                        if leader_store_id.unwrap() != meta.store_id.unwrap() {
-                            continue;
-                        }
-                        let mut read_state = ReadState::default();
-                        if let Some(rrp) = meta.region_read_progress.get(&region_id) {
-                            let rs = rrp.read_state();
-                            read_state.set_applied_index(rs.idx);
-                            read_state.set_safe_ts(rs.ts);
-                        }
-                        for peer in region.get_peers() {
-                            if peer.store_id == store_id && peer.id == *leader_id {
-                                resp_map.entry(region_id).or_default().push(store_id);
-                                continue;
-                            }
-                            if peer.get_role() == PeerRole::Learner {
-                                continue;
-                            }
-                            let mut leader_info = LeaderInfo::default();
-                            leader_info.set_peer_id(*leader_id);
-                            leader_info.set_term(*term);
-                            leader_info.set_region_id(region_id);
-                            leader_info.set_region_epoch(region.get_region_epoch().clone());
-                            leader_info.set_read_state(read_state.clone());
-                            if leader_info_size == 0 {
-                                leader_info_size = leader_info.compute_size() as usize;
-                            }
-                            store_map
-                                .entry(peer.store_id)
-                                .or_default()
-                                .push(leader_info);
-                        }
-                        region_map.insert(region_id, region.clone());
+        // region_id -> `(Vec<Peer>, LeaderInfo)`
+        let info_map = region_read_progress.dump_leader_infos(&regions);
+
+        for (region_id, (peer_list, leader_info)) in info_map {
+            let leader_id = leader_info.get_peer_id();
+            // Check if the leader in this store
+            match find_store_id(&peer_list, leader_id) {
```

Comment:
You mean `find_store_id(&peer_list, leader_id) != Some(store_id)`?

Reply:
Yes